tests/test_skvbc_auto_view_change.py

```diff
@@ -0,0 +1,153 @@
+# Concord
+#
+# Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+import unittest
+
+import os.path
+import random
+import trio
+
+from util import bft
+from util import skvbc as kvbc
+
+KEY_FILE_PREFIX = "replica_keys_"
+
+def start_replica_cmd(builddir, replica_id):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+    Note each arguments is an element in a list.
+    """
+    statusTimerMilli = "500"
+    viewChangeTimeoutMilli = "1000"
+    autoViewChangeTimeoutMilli = "5000"
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", statusTimerMilli,
+            "-v", viewChangeTimeoutMilli,
+            "-a", autoViewChangeTimeoutMilli]
+
+class SkvbcAutoViewChangeTest(unittest.TestCase):
+
+    def test_auto_vc_all_nodes_up_no_requests(self):
+        """
+        This test aims to validate automatic view change
+        in the absence of any client messages:
+        1) Start a full BFT network
+        2) Do nothing (wait for automatic view change to kick-in)
+        3) Check that view change has occurred (necessarily, automatic view change)
+        """
+        trio.run(self._test_auto_vc_all_nodes_up_no_requests)
+
+    async def _test_auto_vc_all_nodes_up_no_requests(self):
+        for bft_config in bft.interesting_configs():
+            config = bft.TestConfig(n=bft_config['n'],
+                                    f=bft_config['f'],
+                                    c=bft_config['c'],
+                                    num_clients=bft_config['num_clients'],
+                                    key_file_prefix=KEY_FILE_PREFIX,
+                                    start_replica_cmd=start_replica_cmd)
+            with bft.BftTestNetwork(config) as bft_network:
+                await bft_network.init()
+                bft_network.start_all_replicas()
+
+                initial_primary = 0
+
+                # do nothing - just wait for an automatic view change
```

Comment:
Regardless of a 10 second wait for a 5 second timeout, we shouldn't be relying on view change to have taken place based on a sleep. We should instead do a wait_for after 5 seconds, retrying every 100ms or something until it passes, and maybe timeout at 10 or 20 seconds. This not only speeds up the test, but if we discover really slow machines timing out during the test, we can bump the wait without making all test runs wait that long.

It looks like `bft_network.get_view_number` already does a wait_for, but it doesn't fail after 10 seconds but moves on after. This is presumably why you needed the assert afterwords also. We should instead fail after 10 seconds. I'd rewrite that function or add a new one and explicitly name it `wait_for_view_number`. I'd also call it after a 5 second sleep to shorten test runs.

Reply:
Hi Andrew, thanks for the detailed review! I have decided to address this comment (along with other related improvements) in a separate PR https://github.com/vmware/concord-bft/pull/320

This way, I can unblock Toly's PR https://github.com/vmware/concord-bft/pull/318, and at the same time keep the focus of the "automatic view change test" PR.